pkg/sources/s3/progress_tracker.go

```diff
@@ -0,0 +1,208 @@
+	if progress == nil {
+		ctx.Logger().Info("Nil progress provided. Progress initialized.")
+		progress = new(sources.Progress)
+	}
+
+	return &ProgressTracker{
+		// We are resuming if we have completed objects from a previous scan.
+		isResuming:       progress.SectionsCompleted > 0 || progress.SectionsRemaining > 0,
+		completedObjects: make([]bool, defaultMaxObjectsPerPage),
+		completionOrder:  make([]int, 0, defaultMaxObjectsPerPage),
+		enabled:          enabled,
+		progress:         progress,
+		baseCompleted:    progress.SectionsCompleted,
+	}
+}
+
+// Reset prepares the tracker for a new page of objects by clearing the completion state.
+func (p *ProgressTracker) Reset(_ context.Context) {
+	if !p.enabled {
+		return
+	}
+
+	p.Lock()
+	defer p.Unlock()
+	// Store the current completed count before moving to next page.
+	p.baseCompleted = p.progress.SectionsCompleted
+	p.currentPageSize = 0
+	p.completedObjects = make([]bool, defaultMaxObjectsPerPage)
+	p.completionOrder = make([]int, 0, defaultMaxObjectsPerPage)
+}
+
+// ResumeInfo represents the state needed to resume an interrupted operation.
+// It contains the necessary information to continue processing from the last
+// successfully processed item.
+type ResumeInfo struct {
+	CurrentBucket string `json:"current_bucket"` // Current bucket being scanned
+	StartAfter    string `json:"start_after"`    // Last processed object key
+}
+
+// GetResumePoint retrieves the last saved checkpoint state if one exists.
+// It returns nil if progress tracking is disabled or no resume state exists.
+// This method decodes the stored resume information and validates it contains
+// the minimum required data to enable resumption.
+func (p *ProgressTracker) GetResumePoint(ctx context.Context) (ResumeInfo, error) {
+	resume := ResumeInfo{}
+	if p.progress == nil {
+		return resume, errors.New("progress is nil, progress is required for resuming")
+	}
+
+	if !p.enabled || p.progress.EncodedResumeInfo == "" {
+		return resume, nil
+	}
+
+	var resumeInfo ResumeInfo
+	if err := json.Unmarshal([]byte(p.progress.EncodedResumeInfo), &resumeInfo); err != nil {
+		return resume, fmt.Errorf("failed to decode resume info: %w", err)
+	}
+
+	if resumeInfo.CurrentBucket == "" {
+		ctx.Logger().V(2).Info("resume info is missing current bucket, resuming from the beginning")
+		return resume, nil
+	}
+
+	return ResumeInfo{
+		CurrentBucket: resumeInfo.CurrentBucket,
+		StartAfter:    resumeInfo.StartAfter,
+	}, nil
+}
+
+// Complete marks the entire scanning operation as finished and clears the resume state.
+// This should only be called once all scanning operations are complete.
+func (p *ProgressTracker) Complete(_ context.Context, message string) error {
+	// Preserve existing progress counters while clearing resume state.
+	p.progress.SetProgressComplete(
+		int(p.progress.SectionsCompleted),
+		int(p.progress.SectionsRemaining),
```

Comment:
Why preserve the counters?

Reply:
This was because `Complete` is only called at the end of the scan. Since progress was tracked at the object level, there’s no need to set these values. I’ll be removing all progress logic. Progress tracking will still happen at the bucket level, but resuming will remain at the object level.